Hw05

VladimirYelkin opened the pull request:
HomeWork5 (spring jdbc)

VladimirYelkin committed:
HomeWork5
- copied template
- admin command - start console h2
- realized AuthorRepositoryJdbc + tests

HomeWork5
- copied template
- admin command - start console h2
- realized AuthorRepositoryJdbc + tests

VladimirYelkin committed:
HomeWork5
- GenreRepositoryJdbc

VladimirYelkin committed:
HomeWork5
- BookRepositoryJdbc realized

VladimirYelkin committed:
HomeWork5
- fixed checkstyle errors

VladimirYelkin committed:
HomeWork5
- fixed

vlad-iv commented on HW05/src/main/java/ru/otus/hw/models/Book.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class Book {
+    private long id;
```
Можно указать Long, т.к. поле опциональное

VladimirYelkin replied:
исправил

vlad-iv commented on HW05/src/main/java/ru/otus/hw/repositories/AuthorRepositoryJdbc.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Optional;
+
+@Repository
+public class AuthorRepositoryJdbc implements AuthorRepository {
```
(опционально) Java рекомендует добавлять название перед интерфейсом
JdbcAuthorRepository

VladimirYelkin replied:
переименовал классы

vlad-iv commented on HW05/src/main/java/ru/otus/hw/repositories/AuthorRepositoryJdbc.java:
```diff
@@ -0,0 +1,45 @@
+    @Override
+    public Optional<Author> findById(long id) {
+        Map<String, Object> params = Collections.singletonMap("id", id);
+        return Optional.ofNullable(namedParameterJdbcOperations.queryForObject(
```
Можно заменить запрос на query
или добавить обработку исключения, когда данные не найдены
так же в остальных методах  findById

vlad-iv commented on HW05/src/main/java/ru/otus/hw/repositories/BookRepositoryJdbc.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Override
+    public Book save(Book book) {
+        if (book.getId() == 0) {
```
== null

VladimirYelkin replied:
исправлено

vlad-iv commented on HW05/src/main/java/ru/otus/hw/repositories/BookRepositoryJdbc.java:
```diff
@@ -0,0 +1,188 @@
+    public void deleteById(long id) {
+        Map<String, Object> params = Collections.singletonMap("id", id);
+        namedParameterJdbcOperations.update("delete from books where id  = :id", params);
+        removeGenresRelationsForBookId(id);
```
предлагаю сначала удалить связи, а потом сущность

VladimirYelkin replied:
исправил

vlad-iv commented on HW05/src/main/java/ru/otus/hw/repositories/BookRepositoryJdbc.java:
```diff
@@ -0,0 +1,188 @@
+    private void mergeBooksInfo(List<Book> booksWithoutGenres, List<Genre> genres,
+                                List<BookGenreRelation> relations) {
+        Map<Long, Genre> genreMap = genres.stream().collect(Collectors.toMap(Genre::getId, Function.identity()));
+        booksWithoutGenres.forEach(book -> {
```
Предлагаю упросить реализацию (текущая сложность O(N^2) ожидается O(N))
Например преобразовать в  хеш-таблицу для книг без жанров
и в цикле по  relations находить книги + добавлять в нее жанр

VladimirYelkin replied:
переделал

vlad-iv commented on HW05/src/main/java/ru/otus/hw/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Override
+    public Book update(long id, String title, long authorId, List<Long> genresIds) {
+        return save(id, title, authorId, genresIds);
```
Осталось проверить, что книга существует.
проверку можно реализовать или в слое сервис
или в репозитории (проверить, что обновили 1 запись)

VladimirYelkin replied:
исправил

vlad-iv commented on HW05/src/main/java/ru/otus/hw/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+        var author = authorRepository.findById(authorId)
+                .orElseThrow(() -> new EntityNotFoundException("Author with id %d not found".formatted(authorId)));
+        var genres = genreRepository.findAllByIds(genresIds);
+        if (isEmpty(genres)) {
```
Предлагаю сначала проверить что isEmpty(genresIds)
после этого получить жанры
и сравнить кол-во
Так же заменить тип с List<Long> genresIds на Set<Long> genresIds

vlad-iv commented on HW05/src/main/resources/data.sql:
```diff
@@ -0,0 +1,14 @@
+insert into authors(full_name)
```
Т.к. скрипты запускаются многократно, поэтому insert лучше заменить на merge

VladimirYelkin replied:
заменил на merge

vlad-iv commented on HW05/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,24 @@
+create table authors (
```
Таблицы могут уже существовать, можно добавить в кажду if not exists

VladimirYelkin committed:
HomeWork5
- fixed Book.java
- fixed long to Long in Book
- fixed in schema.sql
- fixed in AuthorRepositoryJdbc
- fixed in BookRepositoryJdbc + added bookwithoutgenres in dataset
- fixed in BookServiceImpl

VladimirYelkin committed:
HomeWork5
- renaming repositories

VladimirYelkin committed:
HomeWork5
- fix checkstyle

VladimirYelkin committed:
HomeWork5
- fix in dataset (from insert to merge)

vlad-iv commented on HW05/src/main/java/ru/otus/hw/repositories/JdbcBookRepository.java:
```diff
@@ -0,0 +1,187 @@
+    private void mergeBooksInfo(List<Book> booksWithoutGenres, List<Genre> genres,
+                                List<BookGenreRelation> relations) {
+        Map<Long, Genre> genreMap = genres.stream().collect(Collectors.toMap(Genre::getId, Function.identity()));
+        Map<Long, List<Genre>> relationsMap = relations.stream()
```
(опционально) Можно реализовать в BookGenreRelationRowMapper заменить его на экстрактор

vlad-iv commented on HW05/src/main/java/ru/otus/hw/services/BookServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+                .orElseThrow(() -> new EntityNotFoundException("Author with id %d not found".formatted(authorId)));
+        List<Genre> genres = isEmpty(genresIds) ? Collections.emptyList()
+                : genreRepository.findAllByIds(List.copyOf(genresIds));
+        if (isEmpty(genres)) {
```
genres.size() != genresIds.size()

VladimirYelkin committed:
HomeWork5
- fix in BookServiceImpl save();
